### PR TITLE
[Merged by Bors] - calculate flat normals for mesh if missing

### PR DESCRIPTION
**Pull request by jakobhellermann**
If the gltf loader encounters a mesh without normal attributes, it will duplicate the vertex attributes and compute flat normals, as defined by https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/README.md#meshes:

> **Implementation Note**: When normals are not specified, client implementations should calculate flat normals.

image

Helps with #1802

### CODE REVIEW

**Comment by mtsr** [comment]
A future optimization might be to have a flat shading `ShaderDef` that sets the [GLSL *flat* qualifer](https://www.khronos.org/opengl/wiki/Type_Qualifier_(GLSL)#Interpolation_qualifiers) and calculates face normals and sets those on each [provoking vertex](https://www.khronos.org/opengl/wiki/Provoking_Vertex).

**Comment by mtsr** [comment]
@jakobhellermann I think it makes sense to copy the discord message with example numbers, just to give some context.

**Comment by jakobhellermann** [comment]
> I think it makes sense to copy the discord message with example numbers, just to give some context.

When forcing the flight helmet to calculate normals (and thus to duplicate the vertices), the combined vertex buffer length for all meshes grows from `55392` to `284166`.
The damaged helmet goes from `14556` to `46356`.
The SciFi helmet stays at `70074`.

The actual examples of models in the [glTF Sample Models](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0#gltf-20-sample-models) are all *very* small:

<details>
<summary>Expand</summary>

Old length: 10
New length: 24

Old length: 4
New length: 6

Old length: 4
New length: 6

Old length: 4
New length: 6

Old length: 4
New length: 6

Old length: 4
New length: 6

Old length: 4
New length: 6

Old length: 4
New length: 6

Old length: 4
New length: 6

Old length: 4
New length: 6

Old length: 3
New length: 3
</details>

**Comment by jakobhellermann** [comment]
Another alternative proposed by `@Dvd` in https://discord.com/channels/691052431525675048/743663924229963868/827973897240444998 is to calculate the normals in shader:
https://github.com/mrdoob/three.js/blob/master/src/renderers/shaders/ShaderChunk/normal_fragment_begin.glsl.js#L4-L12

Links:
https://github.com/google/filament/issues/2088
https://github.com/KhronosGroup/glTF/issues/1605

**Comment by mtsr** [comment]
> When forcing the flight helmet to calculate normals (and thus to duplicate the vertices), the combined vertex buffer length for all meshes grows from `55392` to `284166`.
> The damaged helmet goes from `14556` to `46356`.
> The SciFi helmet stays at `70074`.

You did mention one other interesting thing: Although the vertex buffer length grows significantly, the index buffer is eliminated. Still not a good trade off, given there's likely more properties per vertex, but it helps.

File: crates/bevy_render/src/mesh/mesh.rs [hunk 1]
```diff
@@ -369,6 +399,73 @@ impl Mesh {
...
+                VertexAttributeValues::Uchar4Norm(vec) => *vec = duplicate(&vec, indices),
+            }
+        }
+    }
+
+    /// Calculates the [`Mesh::ATTRIBUTE_NORMAL`] of a mesh.
+    ///
+    /// Panics if [`Indices`] are set.
+    /// Consider calling [Mesh::duplicate_vertices] or export your mesh with normal attributes.
+    pub fn compute_flat_normals(&mut self) {
```
Comment (julhe) [inline crates/bevy_render/src/mesh/mesh.rs:445]: Normals should also be normalized/unit length.

Comment (jakobhellermann) [reply crates/bevy_render/src/mesh/mesh.rs:445]: Right, normals should be actually normal :smile:

File: crates/bevy_gltf/src/loader.rs [hunk 2]
```diff
@@ -146,6 +146,17 @@ async fn load_gltf<'a, 'b>(
...
 
+            if mesh.attribute(Mesh::ATTRIBUTE_NORMAL).is_none() {
+                let vertex_count_before = mesh.count_vertices();
+                mesh.duplicate_vertices();
+                mesh.compute_flat_normals();
+                let vertex_count_after = mesh.count_vertices();
+
+                if vertex_count_before != vertex_count_after {
+                    bevy_log::warn!("Missing vertex normals in indexed geometry, computing them as flat. Vertex count increased from {} to {}", vertex_count_before, vertex_count_after);
+                }
```
Comment (mockersf) [inline crates/bevy_gltf/src/loader.rs:157]: we should log something even if there wasn't a vertex change

File: crates/bevy_gltf/src/loader.rs [hunk 3]
```diff
@@ -146,6 +146,17 @@ async fn load_gltf<'a, 'b>(
...
             };
 
+            if mesh.attribute(Mesh::ATTRIBUTE_NORMAL).is_none() {
+                let vertex_count_before = mesh.count_vertices();
+                mesh.duplicate_vertices();
+                mesh.compute_flat_normals();
+                let vertex_count_after = mesh.count_vertices();
+
+                if vertex_count_before != vertex_count_after {
+                    bevy_log::warn!("Missing vertex normals in indexed geometry, computing them as flat. Vertex count increased from {} to {}", vertex_count_before, vertex_count_after);
```
Comment (cart) [inline crates/bevy_gltf/src/loader.rs:156]: Calculating normals is a "feature" users might want to use. Warnings must be actionable, but users might not have the ability or inclination to add normals to a model (ex: they bought it on an asset store).

Additionally, some meshes / shaders, might not _need_ normals, and calculating them could be "wasted effort". This is also true for #1803.

I think ideally we detect these cases _only when they are relevant_ (ex: a mesh is missing normals and it is used by a shader that needs normals) and we warn or fail. Then users could somehow opt-in to "mesh transformations" like generating zero-ed out buffers or computing normals. The error/warning could provide a list of ways to resolve the error (ex: fix the mesh manually or opt in to a mesh transformation).

Opting in to these transforms would ideally be a part of the "asset configuration", but bevy_asset currently doesn't support asset config. Distill does (via .meta files), so long term I think thats our best bet.

Its hard to say what we should do in the short term. I'm curious to hear other peoples thoughts. Some ideas:

1. Just warn like we do here and accept that there will be un-actionable warnings for some valid use cases (and open an issue calling this out so we dont forget).
2. Use something like #1665 to allow users to apply the transformation manually to suppress the warning.
3. Add "asset config" to the current bevy_asset.

**Comment by julhe** [comment]
Okay, this all(#1808, #1803) falls down to the PR I made last year to force attributes to be defined. https://github.com/bevyengine/bevy/pull/870 I made this because of the reasons mentiond in the PR, but seeing all the trouble it causes... 😬 

I guess I can try out to re-add the fallback buffer. But this time, missing attributes are defined with the already existing data. This would mean you get garbage vertex data (and I'm not sure how multi platform friendly this is), but at least bevy doesn't crash. Other alternative would be to raise a warning and don't draw the mesh at all/display with error material, Unity style.

**Comment by mockersf** [comment]
it's not exactly the same that providing a fallback buffer. In this case we follow the gltf spec, that is to compute flat normals if missing.

In most case it's better to fix/cleanup the data at its source because you can do so with an idea of its meaning/goal rather than in the buffer when it's "just data"

still a fallback buffer could make sense to avoid crashing, if that's the target we want for Bevy

**Comment by cart** [comment]
Hmm yeah if the gltf spec wants us to automatically calculate flat normals if missing, then I guess we should do that. I don't recall seeing that in the "uber gltf visual png spec doc". Do you have a link to the source of that info?

In the general "missing mesh attribute" case, I _think_ I'd prefer a more explicit "mesh fixup" workflow where developers are notified that their mesh is missing something and they then take some action to apply a "fix" (ex: opt in to zeroed out buffers or add it to the mesh source). This "general case" might be a good candidate for an RFC so we can outline the use cases, desired apis, and prior art.

**Comment by mockersf** [comment]
> Do you have a link to the source of that info?

@jakobhellermann put the link in his description: https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/README.md#meshes:

The part about normal is not directly after but it reads:

> Implementation Note: When normals are not specified, client implementations should calculate flat normals.

**Comment by cart** [comment]
> @jakobhellermann put the link in the pr description: https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/README.md#meshes:

Whoops. Sorry about that :sweat_smile:

**Comment by cart** [comment]
Given that normal calculation is part of the spec, I think we should probably just change the warning to a debug log and be done with it :)

**Comment by jakobhellermann** [comment]
> Given that normal calculation is part of the spec, I think we should probably just change the warning to a debug log and be done with it :)

It's a `debug!` now.

File: crates/bevy_gltf/src/loader.rs (hunk 2 above)
Comment (jakobhellermann) [reply crates/bevy_gltf/src/loader.rs]: Done.

File: crates/bevy_gltf/src/loader.rs [hunk 4]
```diff
@@ -153,6 +153,22 @@ async fn load_gltf<'a, 'b>(
...
+            if mesh.attribute(Mesh::ATTRIBUTE_NORMAL).is_none() {
+                let vertex_count_before = mesh.count_vertices();
+                mesh.duplicate_vertices();
+                mesh.compute_flat_normals();
+                let vertex_count_after = mesh.count_vertices();
+
+                if vertex_count_before != vertex_count_after {
+                    bevy_log::debug!("Missing vertex normals in indexed geometry, computing them as flat. Vertex count increased from {} to {}", vertex_count_before, vertex_count_after);
+                }
+                if vertex_count_before != vertex_count_after {
```
Comment (cart) [inline crates/bevy_gltf/src/loader.rs:165]: looks like we have dupe checks here

**Pull request closed without merging**